### PR TITLE
python3Packages.dm-tree: refactor to build from sources

### PR DESCRIPTION
**Pull request by ndl**
###### Motivation for this change

Part of "extra JAX libraries / frameworks" implementation, see the discussion in https://github.com/NixOS/nixpkgs/pull/152754

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/dm-tree/cmake.patch [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
++find_package(pybind11 CONFIG REQUIRED)
+ 
+ # Needed to disable Abseil tests.
+ set (BUILD_TESTING OFF)
+ 
+ # Include abseil-cpp.
+-set(ABSEIL_VER 20210324.2)
+-include(ExternalProject)
+-ExternalProject_Add(abseil-cpp
+-  GIT_REPOSITORY    https://github.com/abseil/abseil-cpp.git
```
Comment (jonringer) [inline pkgs/development/python-modules/dm-tree/cmake.patch:31]: Use of vendored source is an anti-pattern. Would be nice if upstream supported `find_package`

File: pkgs/development/python-modules/dm-tree/default.nix [hunk 2]
```diff
@@ -1,47 +1,56 @@
...
-  src = fetchPypi {
-    inherit version format;
-    sha256 = "1f71dy5xa5ywa5chbdhpdf8k0w1v9cvpn3qyk8nnjm79j90la9c4";
-    pname = "dm_tree";
-    dist = "cp39";
-    python = "cp39";
-    abi = "cp39";
-    platform = "manylinux_2_24_x86_64";
+  # As of 2021-12-29, the latest stable version still builds with Bazel.
+  version = "42e87fda83278e2eb32bb55225e1d1511e77c10c";
```
Comment (jonringer) [inline pkgs/development/python-modules/dm-tree/default.nix:16]: please use `unstable-YYYY-MM-DD` where the date is the date of the commit used

File: pkgs/development/python-modules/dm-tree/cmake.patch (hunk 1 above)
Comment (jonringer) [reply pkgs/development/python-modules/dm-tree/cmake.patch:31]: not a blocker for this PR though

File: pkgs/development/python-modules/dm-tree/default.nix [hunk 3]
```diff
@@ -1,47 +1,56 @@
...
-, isPy39
-, lib
-, six
-, stdenv
+, abseil-cpp
+, absl-py
+, attrs
+, numpy
+, pybind11
+, wrapt
```
Comment (samuela) [inline pkgs/development/python-modules/dm-tree/default.nix:10]: nit: alphabetize

File: pkgs/development/python-modules/dm-tree/default.nix [hunk 4]
```diff
@@ -1,47 +1,56 @@
...
-    python = "cp39";
-    abi = "cp39";
-    platform = "manylinux_2_24_x86_64";
+  # As of 2021-12-29, the latest stable version still builds with Bazel.
+  version = "42e87fda83278e2eb32bb55225e1d1511e77c10c";
+
+  src = fetchFromGitHub {
+    owner = "deepmind";
+    repo = "tree";
+    rev = "${version}";
```
Comment (samuela) [inline pkgs/development/python-modules/dm-tree/default.nix:21]:
```suggestion
    rev = version;
```

**Review by samuela — COMMENTED** [review]
LGTM after comments are addressed

File: pkgs/development/python-modules/dm-tree/cmake.patch (hunk 1 above)
Comment (ndl) [reply pkgs/development/python-modules/dm-tree/cmake.patch]: Rebased to the most recent commit and changed the patch to hopefully be submittable to the upstream (= falling back to vendored deps if `find_package` fails), will try to open PR with them but keeping the patch here for now as I want to decouple upstream PR from this one.

File: pkgs/development/python-modules/dm-tree/default.nix (hunk 2 above)
Comment (ndl) [reply pkgs/development/python-modules/dm-tree/default.nix]: Done.

File: pkgs/development/python-modules/dm-tree/default.nix (hunk 3 above)
Comment (ndl) [reply pkgs/development/python-modules/dm-tree/default.nix]: Done.

File: pkgs/development/python-modules/dm-tree/default.nix (hunk 4 above)
Comment (ndl) [reply pkgs/development/python-modules/dm-tree/default.nix]: Changed version as per request above => moved revision here as a literal.

**Review by samuela — APPROVED** [review]
LGTM I'll squash and merge tomorrow unless anyone objects.

**Pull request merged**